_posts/2018-05-20-Y-The-Most-Beautiful-Idea-in-Computer-Science.md

```diff
@@ -0,0 +1,611 @@
+---
+layout: post
+title: Y: The Most Beautiful Idea in Computer Science explained in JavaScript
+author: Lucas Fernandes da Costa
+tags: functional-programming combinators javascript y-combinator
+---
+
+In this post we will talk about **one of the most beautiful ideas in computer science: the Y-Combinator**. And no, I'm not talking about [the VC firm in Sillicon Valley](http://www.ycombinator.com/), even though this post will explain why they've got this name.
+
+Put in simple terms, **the Y-Combinator (also known as the fixed-point combinator) is a way of doing recursion in a language that does not explicitly supports it**.
+
+Let's say you want to implement a recursive [factorial function](https://en.wikipedia.org/wiki/Factorial). In JavaScript, for example, you could simply do this:
+
+```js
+const factorial = n => n === 0 ? 1 : n * factorial(n - 1);
```

Comment:
#### Pet peeve warning!

I really don't like the new syntax for functions in JavaScript.

Anyone with basic JavaScript understanding, or experience with practically any other programming language will be able to understand the following.

```js
function factorial(n) {
    if (n === 0) {
        return 1;
    }

    return n * factorial(n - 1);
}
```

Reply:
I agree with you.

I wrote them this way because it resembles more lambda calculus and keep thinks shorter, so it's actually easier to cite certain parts of functions when I'm mentioning them in text. If I had to quote this function in text it would be harder to specify which exact part of it I'm talking about or take too much space to expand all the definitions I'd like to.